src/kvstore/listener/elasticsearch/ESListener.cpp

```diff
@@ -25,48 +25,112 @@ void ESListener::init() {
     }
-    hc.connType = c.conn_type_ref().has_value() ? *c.get_conn_type() : "http";
-    esClients_.emplace_back(std::move(hc));
+    std::string protocol = c.conn_type_ref().has_value() ? *c.get_conn_type() : "http";
+    esClients.emplace_back(HttpClient::instance(), protocol, host.toString(), user, password);
   }
-
+  esAdapter_.setClients(std::move(esClients));
   auto sRet = schemaMan_->toGraphSpaceName(spaceId_);
   if (!sRet.ok()) {
     LOG(FATAL) << "space name error";
   }
   spaceName_ = std::make_unique<std::string>(sRet.value());
 }
 
-bool ESListener::apply(const std::vector<KV>& data) {
-  std::vector<nebula::plugin::DocItem> docItems;
-  for (const auto& kv : data) {
-    if (!nebula::NebulaKeyUtils::isTag(vIdLen_, kv.first) &&
-        !nebula::NebulaKeyUtils::isEdge(vIdLen_, kv.first)) {
-      continue;
+bool ESListener::apply(const BatchHolder& batch) {
+  nebula::plugin::ESBulk bulk;
+  auto callback = [&bulk](BatchLogType type,
+                          const std::string& index,
+                          const std::string& vid,
+                          const std::string& src,
+                          const std::string& dst,
+                          int64_t rank,
+                          const std::string& text) {
+    if (type == BatchLogType::OP_BATCH_PUT) {
+      bulk.put(index, vid, src, dst, rank, text);
+    } else if (type == BatchLogType::OP_BATCH_REMOVE) {
+      bulk.delete_(index, vid, src, dst, rank);
+    } else {
+      LOG(FATAL) << "Unexpect";
     }
-    if (!appendDocItem(docItems, kv)) {
+  };
+  for (const auto& log : batch.getBatch()) {
+    pickTagAndEdgeData(std::get<0>(log), std::get<1>(log), std::get<2>(log), callback);
+  }
+  if (!bulk.empty()) {
+    auto status = esAdapter_.bulk(bulk);
+    if (!status.ok()) {
+      LOG(ERROR) << status;
       return false;
     }
-    if (docItems.size() >= static_cast<size_t>(FLAGS_ft_bulk_batch_size)) {
-      auto suc = writeData(docItems);
-      if (!suc) {
-        return suc;
-      }
-      docItems.clear();
-    }
-  }
-  if (!docItems.empty()) {
-    return writeData(docItems);
   }
   return true;
 }
 
+void ESListener::pickTagAndEdgeData(BatchLogType type,
+                                    const std::string& key,
+                                    const std::string& value,
+                                    const PickFunc& callback) {
+  if (nebula::NebulaKeyUtils::isTag(vIdLen_, key)) {
+    auto tagId = NebulaKeyUtils::getTagId(vIdLen_, key);
+    auto ftIndexRes = schemaMan_->getFTIndex(spaceId_, tagId);
+    if (!ftIndexRes.ok()) {
+      return;
+    }
+    auto ftIndex = std::move(ftIndexRes).value();
+    auto reader = RowReaderWrapper::getTagPropReader(schemaMan_, spaceId_, tagId, value);
+    if (reader == nullptr) {
+      LOG(ERROR) << "get tag reader failed, tagID " << tagId;
+      return;
+    }
+    if (ftIndex.second.get_fields().size() > 1) {
+      LOG(ERROR) << "Only one field will create fulltext index";
```

Comment:
This part of syntax has been modified in Parser, so there can only be one field. In order to minimize changes to IDL, only one check is added here.

This will never be executed.